tools/cldr/cldr-to-icu/README.md

```diff
@@ -40,65 +64,132 @@ relies on a pre-processing step, and the CLDR data must come from the separate
 "staging" repository (i.e. https://github.com/unicode-org/cldr-staging) or be
 pre-processed locally into a different directory.
 
+:point_right: **Note**: the 3 folders can also be overridden:
```

Comment:
optional: do we need three ways to do this?

Reply:
Unfortunately some is legacy.
We have a lot of documentation using 3 folders.
Pointing to the 3 paths where the 3 repos are cloned (icu, cldr, and cldr-staging).

There is also code relaying on that.
The cldr-code expects CLDR_DIR to be set. And not set in environment, but as Java property.
The ant did not allow one to set them from command line, but from the ant script itself, and (in some cases) from environment.
And then it used that set value(s) to set both environment AND the Java property.
Somewhere in the code you will see that I explicityl do a System.setProperty to keep thinks the way ant set them up.

So unfortunately it was tangled pretty badly.
I only kept things as they were.

There is quite a bit of refactoring that I would like to do.
And this can be one of them.
But has to also be coordinated with cldr-code, and a lot of doc.

Reply:
My question is not about the three folders, but about the three ways of defining them:
- as a shell environment variable
- as a JVM property with -Dname=value
- as a command-line option

Reply:
Not a blocker, but maybe worth thinking about?

Reply:
The existing code and doc assumed both shell environment variable and JVM property

The docs say "set this environment variable" everywhere.
And the defined values are used in a lot of copy / paste commands.

Then the code (in several places, including cldr-code, which is not even icu) assumed that some are set as Java properties.
And failed if not.
Ant was "magically" taking some of the environment variables and making them Java properties

For example:
```xml
<condition property="cldrDataDir" value="${env.CLDR_DATA_DIR}">
    <isset property="env.CLDR_DATA_DIR"/>
</condition>
```

And some were "fixed" properties in the ant script:
```xml
<property name="outDir" value="${basedir}/../../../icu4c/source/data/"/>
```

Then the Ant `convert` invoked Ant again as "convert-impl" by defining CLDR_DIR in environment, and passing other things as Java properties :-(
```xml
<exec executable="ant" searchpath="true" failonerror="true">
    <env key="CLDR_DIR" value="${cldrDataDir}" />
    ...
    <arg value="-DoutDir=${outDir}"/>
    ...
```

And finally the real "convert" task (in the "convert-impl" target) took some of them and invoked setters in the associated Java class (extending Task):
```xml
<convert cldrDir="${cldrDir}" outputDir="${outDir}"  ... />
```

But the point is: some of these were both Java properties and environment variables at the same time.
I only preserved the existing behavior.
AND it is only for the folders.

So the Java code never looked at environment vars, only Ant did.
Java looked at (a few) properties + setters.

---

I had a chat with Elango (and a whiteboard) on what to do about it.

The main question is really between JVM properties vs command line options.
I was split between them, each have pros and cons.

In the end CLI won. The rest needs to be unified.
But not a hight priority, and not in this CL.
It will need to (among others) change the cldr-code library in the cldr repo.

Note that only the 3 folders have this "override behavior", not the rest.
Because they had it already.
Grep the old code (before my CL):
```sh
$ grep -r   System.getProperty src/

src//test/java/org/unicode/icu/tool/cldrtoicu/SupplementalDataTest.java:        Path cldrRoot = Paths.get(System.getProperty("CLDR_DIR"));
src//main/java/org/unicode/icu/tool/cldrtoicu/IcuConverterConfig.java:        Optional.ofNullable(System.getProperty("ICU_DIR", null))
src//main/java/org/unicode/icu/tool/cldrtoicu/DebugWriter.java:        String cldrPath = System.getProperty("CLDR_DIR", System.getenv("CLDR_DIR"));
```

And note this in `DebugWriter.java`:
```java
String cldrPath = System.getProperty("CLDR_DIR", System.getenv("CLDR_DIR"));
```

This assumption was there.
And it was too much surgery to fix in one step.
That's why I was forced to do a `System.setProperty("CLDR_DIR", cldrDir);` (in `Cldr2IcuCliOptions.java`)

---

TLDR: I am not arguing that it is good, or that I like it.
What I am trying to say is that was unclean before, it is the same now.

Deserves a cleanup, but it is not mavenization, and is not as simple as removing this override that I implemented here.

For example can't remove it from cldr-code, because CLDR itself also depends on it.
So the processes / scripts / docs in CLDR would need to be updated BEFORE removing it from this code.